torchdata/nodes/samplers/multi_node_weighted_sampler.py

```diff
@@ -0,0 +1,224 @@
+
+    def reset(self, initial_state: Optional[Dict[str, Any]] = None):
+        super().reset(initial_state)
+
+        if initial_state is not None:
+            self._num_yielded = initial_state[self.NUM_YIELDED_KEY]
+            self._weighted_sampler = _WeightedSampler(
+                weights=self.weights,
+                seed=self.seed,
+                rank=self.rank,
+                world_size=self.world_size,
+                initial_state=initial_state[self.WEIGHTED_SAMPLER_STATE_KEY],
+            )
+            self._datasets_exhausted = initial_state[self.DATASETS_EXHAUSTED_KEY]
+            for k in self.dataset_names:
+                self.source_nodes[k].reset(initial_state[self.DATASET_NODE_STATES_KEY][k])
+        else:
+            # Force a fresh iterator from all source nodes
+            self._num_yielded = 0
+            self._weighted_sampler = _WeightedSampler(
+                weights=self.weights,
+                seed=self.seed,
+                rank=self.rank,
+                world_size=self.world_size,
+            )
+            self._datasets_exhausted = {key: False for key in self.weights.keys()}
+            for k in self.dataset_names:
+                self.source_nodes[k].reset()
+
+    def _check_for_stop_iteration(self) -> None:
+        if all(self._datasets_exhausted.values()):
+            # Raise StopIteration if all datasets are exhausted,
+            # this covers for both ALL_DATASETS_EXHAUSTED and CYCLE_UNTIL_ALL_DATASETS_EXHAUSTED
+            raise StopIteration()
+
+        # Raise StopIteration is StopCriteria is FIRST_DATASET_EXHAUSTED and
+        # the first dataset is exhausted. Doing this to correctly catch StopIteration
+        # when trying next(it) on already exhausted iterator
+        if self.stop_criteria == StopCriteria.FIRST_DATASET_EXHAUSTED and any(self._datasets_exhausted.values()):
+            raise StopIteration()
+
+        return
+
+    def next(self) -> T:
+        while True:
+            self._check_for_stop_iteration()
+
+            # Fetch the next item's key from the weighted sampler
+            key = next(self._weighted_sampler)
+            try:
+                if self._datasets_exhausted[key] and self.stop_criteria == StopCriteria.ALL_DATASETS_EXHAUSTED:
+                    # Before fetching a new item check if key corresponds to an already
+                    # exhaused dataset and StopCriteria is ALL_DATASETS_EXHAUSTED, move to next key
+                    # return next(self)  # omit recursive call
+                    continue
+                item = next(self.source_nodes[key])
+            except StopIteration:
+                # Mark the dataset as exhausted
+                self._datasets_exhausted[key] = True
+
+                # Based on updated _check_for_stop_iteration, check if we should raise StopIteration
+                self._check_for_stop_iteration()
+
+                # If StopCriteria is ALL_DATASETS_EXHAUSTED, move to next key
```

Comment:
Let's describe the else condition here (ie we should only hit "else" if "cycle_until_all_datsets_exhausted" and in that case we need to reset). Wondering if you think it would be more clear to frame the condition that way: ie 
```
if CYCLE_UNTIL_ALL_DATASESTS_EXHAUSTED:
  node.reset()
  item = next(node)
  break
else:
  continue
```